File: Python/scheduler.py
#!/usr/bin/env python3
"""
Scheduler — called by brain_loop after any user action.
Proposes specific times (never asks "when?").
Returns schedule proposal → brain_loop writes to schedule.json
"""
import json
import datetime
from pathlib import Path
from scorer import score_all, ENERGY_CURVE

BASE          = Path.home() / "notchly/v2"
GCAL_CACHE    = BASE / "cache" / "gcal_cache.json"
SCHEDULE_FILE = BASE / "schedule.json"


def write_atomic(path: Path, data) -> None:
    import os
    tmp = path.with_suffix(".tmp")
    with open(tmp, "w") as f:
        json.dump(data, f, indent=2, default=str)
    os.rename(tmp, path)


def load_today_events() -> list:
    try:
        with open(GCAL_CACHE) as f:
            data = json.load(f)
        return data.get("today", [])
    except Exception:
        return []


def free_blocks(events: list) -> list:
    """Compute free time blocks today (>= 15min)"""
    now = datetime.datetime.now()
    day_end = now.replace(hour=23, minute=0, second=0)
    cursor = now

    blocks = []
    for event in sorted(events, key=lambda e: e.get("time", "")):
        try:
            parts = event["time"].split(":")
            ev_start = now.replace(hour=int(parts[0]), minute=int(parts[1]))
            dur_str = event.get("duration", "0m")
            dur_min = 0
            if "h" in dur_str:
                h, rest = dur_str.split("h")
                dur_min += int(h.strip()) * 60
                if "m" in rest:
                    dur_min += int(rest.replace("m", "").strip() or "0")
            elif "m" in dur_str:
                dur_min = int(dur_str.replace("m", "").strip())
            ev_end = ev_start + datetime.timedelta(minutes=dur_min)

            if ev_start > cursor:
                gap_min = (ev_start - cursor).total_seconds() / 60
                if gap_min >= 15:
                    blocks.append({"start": cursor, "end": ev_start, "minutes": int(gap_min)})
            cursor = max(cursor, ev_end)
        except Exception:
            continue

    if cursor < day_end:
        gap_min = (day_end - cursor).total_seconds() / 60
        if gap_min >= 15:
            blocks.append({"start": cursor, "end": day_end, "minutes": int(gap_min)})
    return blocks


HIGH_ENERGY_CATS  = {"deep_work", "creative", "study"}
LOW_ENERGY_CATS   = {"admin", "review"}
SKIP_CATS         = {"class"}


def match_tasks_to_slots(tasks: list, blocks: list, context: dict, profile: dict) -> list:
    """Match tasks to free blocks, respecting energy levels"""
    schedule = []
    assigned = set()

    for block in blocks:
        hour = block["start"].hour
        if profile and "energy_by_hour" in profile:
            energy = profile["energy_by_hour"].get(str(hour), ENERGY_CURVE.get(hour, 5.0))
        else:
            energy = ENERGY_CURVE.get(hour, 5.0)

        is_peak = energy >= 8
        is_dip  = energy <= 5

        for task in tasks:
            tid = task.get("id", task.get("title", ""))
            if tid in assigned:
                continue
            cat = task.get("category", "other")
            if cat in SKIP_CATS:
                continue
            if is_peak and cat not in HIGH_ENERGY_CATS and cat not in {"meeting", "exercise"}:
                continue
            if is_dip and cat not in LOW_ENERGY_CATS and cat not in {"meal", "break", "other"}:
                continue

            est_min = task.get("estimatedMinutes", 30)
            if block["minutes"] < est_min:
                continue

            # Add 10min context switch buffer
            start = block["start"] + datetime.timedelta(minutes=10)
            end   = start + datetime.timedelta(minutes=est_min)
            if end > block["end"]:
                continue

            schedule.append({
                "taskID": tid,
                "taskTitle": task.get("title", ""),
                "start": start.isoformat(),
                "end": end.isoformat(),
                "energy": energy,
            })
            assigned.add(tid)

            # Update block cursor
            block["start"] = end
            block["minutes"] = int((block["end"] - end).total_seconds() / 60)
            break

    return schedule


def propose_schedule(wm: dict, context: dict, profile: dict) -> list:
    tasks  = wm.get("taskQueue", [])
    scored = score_all(tasks, context, profile)
    events = load_today_events()
    blocks = free_blocks(events)
    return match_tasks_to_slots(scored, blocks, context, profile)


def run_once(wm: dict, context: dict, profile: dict) -> None:
    schedule = propose_schedule(wm, context, profile)
    write_atomic(SCHEDULE_FILE, {"schedule": schedule, "generated_at": datetime.datetime.now().isoformat()})
    print(f"[scheduler] proposed {len(schedule)} scheduled blocks")

File: Sources/Brain/Scheduler.swift
import Foundation
import EventKit

// Dynamic rescheduler — called after EVERY user action.
// AI never asks "when?". AI proposes specific times.
class Scheduler {
    static let shared = Scheduler()
    private init() {}

    // Called after: done/skip/postpone/extend/swipe
    func rescheduleAll(tasks: inout [NotchTask], context: ContextSnapshot) {
        // 1. Score all incomplete tasks
        tasks = PriorityScorer.shared.scoreAll(
            tasks.filter { !$0.isCompleted }, context: context)

        // 2. Filter blocked (C=0 = wrong context)
        let available = tasks.filter { $0.contextFit > 0 }

        // 3. Tasks already sorted by P score from scorer
        // 4. Get today's free blocks
        let freeBlocks = CalendarReader.shared.todayFreeBlocks()

        // 5. Match tasks to slots
        var schedule: [ScheduledBlock] = []
        var usedTime = Date()

        for block in freeBlocks {
            let energy = EnergyModel.shared.currentEnergy(
                at: Calendar.current.component(.hour, from: block.start))
            let isPeak = energy >= 8
            let isDip  = energy <= 5

            // Pick appropriate task category for slot
            let candidates = available.filter { task in
                if isPeak { return [.deepWork, .creative, .study].contains(task.category) }
                if isDip  { return [.admin, .review].contains(task.category) }
                return true
            }

            for task in candidates {
                let duration = TimeInterval(task.estimatedMinutes * 60)
                let slotAvail = block.end.timeIntervalSince(block.start)
                guard slotAvail >= duration else { continue }

                let start = max(block.start, usedTime + 600) // 10min context switch
                let end   = start.addingTimeInterval(duration)
                guard end <= block.end else { continue }

                schedule.append(ScheduledBlock(taskID: task.id, start: start, end: end))
                usedTime = end
                break
            }
        }

        // 6. Apply transit buffer: 15min after class, 20min before first task
        applyBuffers(&schedule)

        // 9. Write schedule atomically
        if let data = try? JSONEncoder().encode(schedule) {
            try? data.write(to: DirectorySetup.schedule, options: .atomic)
        }
    }

    private func applyBuffers(_ schedule: inout [ScheduledBlock]) {
        // Simplified — full buffer logic in Phase G Python scheduler
        // Just ensure 10min gaps between tasks
        for i in 1..<schedule.count {
            let prev = schedule[i-1]
            if schedule[i].start < prev.end + 600 {
                let gap = prev.end + 600
                let duration = schedule[i].end.timeIntervalSince(schedule[i].start)
                schedule[i].start = gap
                schedule[i].end   = gap + duration
            }
        }
    }

    // Auto-reschedule after 3× postpone
    func autoRescheduleTomorrow(_ task: inout NotchTask) {
        let tomorrow = Calendar.current.date(byAdding: .day, value: 1, to: Date())!
        var comps = Calendar.current.dateComponents([.year, .month, .day], from: tomorrow)
        comps.hour   = 10
        comps.minute = 0
        task.scheduledStart = Calendar.current.date(from: comps)
        task.postponeCount  = 0
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// MARK: - Rust Pulse Bridge (WorldState reader)
// Reads ~/notchly/v2/bridge/world_state_canonical.json written by notchly_pulse
// ─────────────────────────────────────────────────────────────────────────────

struct WorldState: Decodable {
    let pulseVersion: String
    let generatedAt: Double
    let aiStatus: String
    let taskQueue: [PulseTask]
    let currentTask: PulseTask?
    let nextTask: PulseTask?
    let missedCount: Int

    enum CodingKeys: String, CodingKey {
        case pulseVersion = "pulse_version"
        case generatedAt  = "generated_at"
        case aiStatus     = "ai_status"
        case taskQueue    = "task_queue"
        case currentTask  = "current_task"
        case nextTask     = "next_task"
        case missedCount  = "missed_count"
    }
}

struct PulseTask: Decodable, Identifiable {
    let id: String
    let title: String
    let subtitle: String?
    let taskType: String
    let durationMinutes: Int
    let progressPct: Double
    let priority: Double
    let appBundle: String?
    let isDone: Bool
    let isSkipped: Bool

    enum CodingKeys: String, CodingKey {
        case id, title, subtitle, priority
        case taskType        = "type"
        case durationMinutes = "duration_minutes"
        case progressPct     = "progress_pct"
        case appBundle       = "app_bundle"
        case isDone          = "is_done"
        case isSkipped       = "is_skipped"
    }

    func toNotchTask() -> NotchTask {
        let cat: TaskCategory = {
            switch taskType {
            case "meal":      return .meal
            case "class":     return .class
            case "exercise":  return .exercise
            case "break":     return .break
            case "deep_work": return .deepWork
            case "study":     return .study
            default:          return .other
            }
        }()
        var t = NotchTask(title: title)
        t.subtitle           = subtitle ?? ""
        t.category           = cat
        t.estimatedMinutes   = durationMinutes
        t.progressPercent    = progressPct
        t.pFinal             = priority
        t.relatedAppBundleID = appBundle
        t.isCompleted        = isDone
        return t
    }
}

final class WorldStateReader {
    static let shared = WorldStateReader()

    private let stateURL = FileManager.default.homeDirectoryForCurrentUser
        .appendingPathComponent("notchly/v2/bridge/world_state_canonical.json")
    private var source: DispatchSourceFileSystemObject?
    private var pollTimer: Timer?
    private var lastTS: Double = 0

    private init() {}

    func start() {
        ensureDirs()
        startFileSource()
        pollTimer = Timer.scheduledTimer(withTimeInterval: 2.0, repeats: true) { [weak self] _ in
            self?.reload()
        }
        reload()
    }

    func stop() {
        source?.cancel()
        pollTimer?.invalidate()
    }

    private func ensureDirs() {
        let base = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("notchly/v2")
        for d in ["bridge", "alerts", "alerts/processed", "queue", "memory"] {
            try? FileManager.default.createDirectory(
                at: base.appendingPathComponent(d),
                withIntermediateDirectories: true)
        }
    }

    private func startFileSource() {
        let dir = stateURL.deletingLastPathComponent()
        let fd = open(dir.path, O_EVTONLY)
        guard fd >= 0 else { return }
        source = DispatchSource.makeFileSystemObjectSource(
            fileDescriptor: fd,
            eventMask: [.write, .rename],
            queue: .global(qos: .utility)
        )
        source?.setEventHandler { [weak self] in self?.reload() }
        source?.setCancelHandler { close(fd) }
        source?.resume()
    }

    private func reload() {
        guard let data = try? Data(contentsOf: stateURL),
              let world = try? JSONDecoder().decode(WorldState.self, from: data),
              world.generatedAt > lastTS else { return }
        lastTS = world.generatedAt
        DispatchQueue.main.async {
            NotchState.shared.applyWorldState(world)
        }
    }
}
